crnlib/crn_find_files.cpp

```diff
@@ -206,22 +206,39 @@ bool find_files::find_internal(const char* pBasepath, const char* pRelpath, cons
     if ((strcmp(ep->d_name, ".") == 0) || (strcmp(ep->d_name, "..") == 0))
       continue;
 
-    bool is_directory = 0;
-    bool is_file = 0;
+    bool is_directory = false;
+    bool is_file = false;
+    bool guessed = false;
 
-    if ( ep->d_type != 0 ) {
+    /* This is the faster guess as it doesn't require any
+    extra IO as everything is already read. */
+    if (ep->d_type != 0) {
       is_directory = (ep->d_type & DT_DIR) != 0;
       is_file = (ep->d_type & DT_REG) != 0;
-    } else {
+      guessed = true;
+    }
+
+    /* Not all filesystems implement d_type which is optional,

```

Comment:
Many filesystems are expected to be affected.

The `dirent.d_type` field is fully optional and one cannot expect it to be implemented. So one should assume the filesystem doesn't support it by default.

Quotes from the `readdir` manpage:

```
 struct dirent {
    ino_t          d_ino;       /* Inode number */
    off_t          d_off;       /* Not an offset; see below */
    unsigned short d_reclen;    /* Length of this record */
    unsigned char  d_type;      /* Type of file; not supported
                                   by all filesystem types */
    char           d_name[256]; /* Null-terminated filename */
 };
```


```
d_type This field contains a value indicating the file type, making it
       possible to avoid the expense of calling lstat(2) if further
       actions depend on the type of the file.

       When a suitable feature test macro is defined (_DEFAULT_SOURCE
       since glibc  2.19,  or  _BSD_SOURCE  on glibc 2.19 and earlier),
       glibc defines the following macro constants for the value returned
       in d_type:

       DT_BLK      This is a block device.

       DT_CHR      This is a character device.

       DT_DIR      This is a directory.

       DT_FIFO     This is a named pipe (FIFO).

       DT_LNK      This is a symbolic link.

       DT_REG      This is a regular file.

       DT_SOCK     This is a UNIX domain socket.

       DT_UNKNOWN  The file type could not be determined.

       Currently,  only  some filesystems (among them: Btrfs, ext2, ext3,
       and ext4) have full support for returning the file type in d_type.
       All applications must properly handle a return of DT_UNKNOWN.
```

Applications querying a file type must support `stat` and may optionally support `dirent.d_type` for faster IO in case the filesystem implements it.

Filesystems implementing `dirent.d_type` are usually the native files systems of the operating system, to bring the maximum performance to the operating system's own software.

Reply:
This makes me thing we should better do `if (ep->d_type != DT_UNKNOWN)`  instead of `if (ep->d_type != 0)`, as `0` is `DT_UNKNOWN`. It would be more explicit. Now done.

Reply:
This was quoting glibc's `readdir` manpage. The macOS man page on my macbox doesn't mention `d_type`, though it looks like HFS+ implements it (the `d_type` check works outside of NFS).

The FreeBSD man page says:

```
STANDARDS
     The closedir(), dirfd(), fdopendir(), opendir(), readdir(), readdir_r(),
     rewinddir(), seekdir() and telldir() functions are expected to conform to
     IEEE Std 1003.1-2008 (“POSIX.1”).  The fdclosedir() function and the
     d_off, d_reclen and d_type fields of struct dirent are non-standard, and
     should not be used in portable programs.
```

So glibc says:

> Currently, only some filesystems (among them: Btrfs, ext2, ext3, and ext4) have full support for returning the file type in d_type. All applications must properly handle a return of DT_UNKNOWN.

And FreeBSD says:

> d_type [field] of struct dirent [is] non-standard, and should not be used in portable programs.

Crunch was non-standard and not portable until I added the `stat` implementation in #41, it just happened that I did a mistake at the time, passing the file basename instead of the file path. This is fixing and finishing the portable and standard implementation.